File: 程式設計：資料結構答案檔/Python/203.py
class TreeNode:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None

    def preorder_traversal(self):
        result = []
        result.append(self.value)

        if self.left:
            result.extend(self.left.preorder_traversal())

        if self.right:
            result.extend(self.right.preorder_traversal())

        return result

def insert_node(root, value):
    if root is None:
        return TreeNode(value)
    else:
        if value < root.value:
            root.left = insert_node(root.left, value)
        else:
            root.right = insert_node(root.right, value)
    return root

def level_order_traversal(root):
    result = []
    if root is None:
        return result

    queue = [root]

    while queue:
        current_node = queue.pop(0)
        result.append(current_node.value)

        if current_node.left:
            queue.append(current_node.left)
        if current_node.right:
            queue.append(current_node.right)

    return result

n = int(input())
values_list = list(map(int, input().split()))

root_node = None
for value in values_list:
    root_node = insert_node(root_node, value)

result_list = level_order_traversal(root_node)
print("Level-order:", end="")
print(" ".join([str(value) for value in result_list]))
print("Preorder:", end="")
preorder_result = root_node.preorder_traversal()
print(" ".join([str(value) for value in preorder_result]))

File: 程式設計：資料結構答案檔/C#/203.cs
using System;
using System.Collections.Generic;

class TreeNode {
    public int Value { get; set; }
    public TreeNode Left { get; set; }
    public TreeNode Right { get; set; }

    public TreeNode(int value) {
        Value = value;
        Left = null;
        Right = null;
    }
}

class BinaryTree {
    public TreeNode Root { get; set; }

    public void InsertNode(int value) {
        Root = InsertNode(Root, value);
    }

    private TreeNode InsertNode(TreeNode root, int value) {
        if (root == null) {
            root = new TreeNode(value);
        } else {
            if (value < root.Value) {
                root.Left = InsertNode(root.Left, value);
            } else {
                root.Right = InsertNode(root.Right, value);
            }
        }
        return root;
    }

    public void LevelOrderTraversal() {
        if (Root == null) return;
        Queue<TreeNode> queue = new Queue<TreeNode>();
        queue.Enqueue(Root);

        bool isFirst = true;

        while (queue.Count > 0) {
            TreeNode currentNode = queue.Dequeue();
            if (!isFirst) Console.Write(" ");
            Console.Write(currentNode.Value);
            isFirst = false;

            if (currentNode.Left != null) queue.Enqueue(currentNode.Left);
            if (currentNode.Right != null) queue.Enqueue(currentNode.Right);
        }
    }

    public void PreorderTraversal() {
        PreorderTraversal(Root, true);
    }

    private void PreorderTraversal(TreeNode root, bool isFirst) {
        if (root == null) return;
        if (!isFirst) Console.Write(" ");
        Console.Write(root.Value);
        PreorderTraversal(root.Left, false);
        PreorderTraversal(root.Right, false);
    }
}

class Program {
    static void Main() {
        int n = int.Parse(Console.ReadLine());
        string[] numbers = Console.ReadLine().Split(' ');
        BinaryTree tree = new BinaryTree();

        foreach (string number in numbers) {
            tree.InsertNode(int.Parse(number));
        }

        Console.Write("Level-order:");
        tree.LevelOrderTraversal();

        Console.Write("\nPreorder:");
        tree.PreorderTraversal();
    }
}
